---
title: '1.Henry_2018 (Healthy Tissue): normalization and rescalingÂ¨'
author: "Romuald Parmentier"
date: "2024-10-24"
output: html_document
---

```{r Load libraries and create output folder, message = F}

library(Seurat)
library(SingleCellExperiment)
library(scran)
library(scater)
library(ggplot2)
library(AnnotationHub)
library(dplyr)

# Functions and palettes
source(paste0(github_dir, "/Dolgos_Custom_Functions.R"))
source(paste0(github_dir, "/Dolgos_Custom_Color_Palettes.R"))

# Create output path
out_path = create_exp_folder(
  github_dir = github_dir,
  samples_ID = "Henry_Tissue",
  exp = "1_normalization_rescaling"
)

```

# Prepare the data

```{r Load files}

raw_data_dir = "/scicore/home/wykopa75/GROUP/rparmentier/sc_RNAseq/Projects/data/Henry_2018/"

# Load count matrix
counts = data.table::fread(file = "/scicore/home/wykopa75/GROUP/rparmentier/sc_RNAseq/Projects/data/Henry_2018/exprMatrix.tsv.gz")

# Load metadata
metadata = read.table(file = "/scicore/home/wykopa75/GROUP/rparmentier/sc_RNAseq/Projects/data/Henry_2018/meta.tsv",
                  fill = T, 
                  header = T, 
                  sep = '\t')

# Store and correct gene names, removing everything before |
genes = counts$gene
genes = sub(pattern = "^.*\\|", replacement = "", genes)

# Rename the rows according new gene names
cells = colnames(counts[,-1])
cells = gsub(x = cells,pattern = "-1-", replacement = "_" )

# Remove gene names from the matrix
counts = counts[,-1]
counts = as.matrix(counts)

rownames(counts) = NULL
colnames(counts) = NULL

```

```{r Transform to sce combined object}

# Combine the slots to create the combined sce object
sce_comb <- SingleCellExperiment( 
  assays = list(
    counts = counts),  
  colData = metadata,
  rowData = DataFrame("SYMBOL" = genes)
) 

rownames(sce_comb) = genes
colnames(sce_comb) = cells

# Change batch number to Sample_Description
sce_comb$Sample_Description = paste0("batch_", sce_comb$batch)

```

```{r Split the sce into a list of sce}

list_sce = list(
  sce_comb[,which(sce_comb$Sample_Description == "batch_1")],
  sce_comb[,which(sce_comb$Sample_Description == "batch_2")],
  sce_comb[,which(sce_comb$Sample_Description == "batch_3")],
  sce_comb[,which(sce_comb$Sample_Description == "batch_4")],
  sce_comb[,which(sce_comb$Sample_Description == "batch_5")],
  sce_comb[,which(sce_comb$Sample_Description == "batch_6")]
  )

names(list_sce) = paste0("batch_",unique(sce_comb$batch))

```

```{r Convert ENSEMBL <--> SYMBOL and add other gene informations}

## Load the annotation resource
EnsDb.Hsapiens.v104 <- AnnotationHub(localHub = TRUE)[["AH95744"]]

sce_id = 1

for(sce in list_sce){
  
  name = names(list_sce)[[sce_id]]
  
  # Extract rowData slot of the first object (same evrywhere)
  rowData_df = as_tibble(rowData(sce))
  
  # Map ENSEMBL ID to SYMBOL
  SYMBOL_to_ENS <- mapIds(
    EnsDb.Hsapiens.v104, 
    keys = rowData_df$SYMBOL, # List against which we want a match
    keytype = "SYMBOL",# Type of the key (predefined list, see help)
    column = "GENEID")
  
  # Add symbol column to rowData_df
  rowData_df$ENSEMBL = SYMBOL_to_ENS
  
  # Filter out NA in ENSEMBL otherwise mapping is not possible
  rowData_df = rowData_df%>%
    dplyr::filter(!is.na(ENSEMBL))
  
  # Map feature type to rowData_df
  ENS_to_BIOTYPE <- mapIds(
    EnsDb.Hsapiens.v104, 
    keys = rowData_df$ENSEMBL, # List against which we want a match
    keytype = "GENEID",# Type of the key (predefined list, see help)
    column = "GENEBIOTYPE") # The matching request
  
  # Add type of feature column to rowData_df
  rowData_df$GENEBIOTYPE = ENS_to_BIOTYPE
  
  # Map chromosome to rowData_df
  ENS_to_chr <- mapIds(
    EnsDb.Hsapiens.v104, 
    keys = rowData_df$ENSEMBL, # List against which we want a match
    keytype = "GENEID",# Type of the key (predefined list, see help)
    column = "SEQNAME") # The matching request
  
  # Add chromosome to rowData_df
  rowData_df$SEQNAME = ENS_to_chr
  
  new_rowData_df = rowData_df %>%
    dplyr::filter(SEQNAME %in% c(1:22, "MT","X","Y")) %>%
    dplyr::filter(GENEBIOTYPE %in% c("protein_coding","lncRNA")) %>%
    dplyr::filter(SYMBOL != "") %>%
    dplyr::select(ENSEMBL,SYMBOL,GENEBIOTYPE,SEQNAME)
  
  # Find common genes between the current SCE object's rownames and the ENSEMBL IDs in new_rowData
  common_genes <- intersect(rownames(sce), new_rowData_df$SYMBOL)
  ind_common_genes = which(rownames(sce) %in% common_genes)
  
  # Subset the counts matrix to include only the common genes found in both the SCE object and new_rowData
  subset_counts <- counts(sce)[ind_common_genes, ]
  
  # Ensure rownames of the subset_counts matrix are updated to the corresponding SYMBOL names from new_rowData
  rownames(subset_counts) <- new_rowData_df$SYMBOL
  
  # Create a new SCE object with:
  # - Subsetted counts matrix (updated gene names)
  # - Updated rowData (with SYMBOLs)
  # - Same colData as the original SCE object
  sce <- SingleCellExperiment(
    assays = list(counts = subset_counts),  # Updated counts matrix
    rowData = DataFrame(new_rowData_df),   # Updated rowData with SYMBOLs
    colData = colData(sce)  # Retain original colData unchanged
  )
  
  list_sce[[name]] <- sce
  
  sce_id = sce_id + 1
  
}

```

```{r Clean the environement to release memory}

rm("sce","EnsDb.Hsapiens.v104")
gc()

```

# Normalization 

No need to remove identified bad quality cells as they were already absent fromn the downloaded object

```{r quickCluster and logNormCounts}

# Initialize ID
sce_id = 1

for (sce in list_sce) {

  set.seed(100)
  
  # Pre-clustering step
  print(paste("Pre-clustering for", names(list_sce)[sce_id], "ongoing..."))
  clust.sce = quickCluster(sce)
  
  # # Adds a sizeFactor col in metadata, sizeFactors are here computed "cluster-wise"
  print(paste("Calculating size factors for", names(list_sce)[sce_id], "ongoing..."))
  new_sce <- computeSumFactors(sce, clusters = clust.sce)
  
  ## Normalize (using already calculated size factors)
  print(paste("Log normalization for", names(list_sce)[sce_id], "ongoing..."))
  new_sce <- logNormCounts(new_sce, size.factors = new_sce$sizeFactor) 
  
  print(paste("Adding", names(list_sce)[sce_id], "to the list_sce object"))
  
  # Update the list_sce
  list_sce[[sce_id]] = new_sce
  
  # Append sce_id
  sce_id = sce_id + 1

}

```

# Batch rescaling 

```{r Rescaling list of sce}

list_sce_rescaled <- batchelor::multiBatchNorm(
  min.mean = 0.1, 
  unlist(list_sce)
)

```


```{r Create a combined rescaled sce object}

# Get total raw counts, first in a list, then combine them into a matrix
counts_comb = lapply(list_sce_rescaled, function(x) {counts(x)})
counts_comb <- Reduce(function(x, y) cbind(x, y), counts_comb)

## Then get total logNormed counts, first in a list, then combine them into a matrix
logcounts_comb = lapply(list_sce_rescaled, function(x) {logcounts(x)})
logcounts_comb <- Reduce(function(x, y) cbind(x, y), logcounts_comb)

# Combine all cell metadata (columns are already the same for each object of the list)
colData_comb = lapply(list_sce_rescaled, function(x) {colData(x)})
colData_comb = Reduce(function(x, y) rbind(x, y), colData_comb)

# Combine the slots to create the combined sce object
sce_comb_rescaled <- SingleCellExperiment( 
  assays = list(counts = counts_comb, logcounts = logcounts_comb),  
  rowData = rowData(list_sce_rescaled[[1]]), # RowData is similar among all elements 
  colData = colData_comb)
 
```


# Export files 

- Rescaled object are intended for further analysis comprising this batch only.
- Non-rescaled object will be used when integrated with other datasets, where re-scaling will be applied among all datasets.

```{r Export RDS files}

# Export list_sce 
saveRDS(
  object = list_sce_rescaled,
  file = paste0(out_path,time_stamp(), "list_sce_normalized.rds"))

# Export re-scaled list_sce 
saveRDS(
  object = list_sce_rescaled,
  file = paste0(out_path,time_stamp(), "list_sce_normalized_rescaled.rds"))

# Export sce_comb object
saveRDS(
  object = sce_comb_rescaled,
  file = paste0(out_path,time_stamp(),"sce_comb_normalized_rescaled.rds"))

```